### PR TITLE
[skip ci] Datasource/jndi dependency race condition - WIP

### PR DESCRIPTION
**Pull request by jamesnetherton**
@tdiesler please can you review this WIP...

My latest attempt at resolving the Spring JNDI race condition documented in #2239.

In summary, this commit makes the following changes:

- Decouples the parsing of the Spring context XML and the subsequent refresh the Spring `ApplicationContext`. This is necessary so that we can figure out required JNDI bindings prior to starting the Spring app.

- Delegates starting of `SpringCamelContext` to a service which has dependencies on all JNDI bindings configured in the Spring `BeanFactory` via `JndiObjectFactoryBean`. 

Some caveats for this to work properly:

- JNDI bindings can only be discovered via `JndiObjectFactoryBean` bean definitions in XML configuration. Using annotation driven configuration via `@Bean` or `@Configuration` will not work. 

- The `JndiObjectFactoryBean` `jndiName` property must not use property placeholders, since these cannot be resolved until `ApplicationContext.refresh()` is called.

In short, the approved means for a Spring application to establish a dependency on some JNDI binding will be to do one of the following in Spring XML configuration:

**Explicit JndiObjectFactoryBean bean**
```xml
<bean id="transactionManager" class="org.springframework.jndi.JndiObjectFactoryBean">
    <property name="jndiName" value="java:/TransactionManager"/>
</bean>
```

**Shortened jee: namespace lookup**
```xml
<jee:jndi-lookup id="userTransaction" jndi-name="java:/jboss/UserTransaction"/>
```

### CODE REVIEW

**Comment by jamesnetherton** [comment]
ok to test

**Comment by jamesnetherton** [comment]
retest this please
